statistics.py initial setup.

sharonyang opened the pull request.

sharonyang committed:
Established basic structure for Statistics.

This includes subclasses and fields. Functions and events have not
been written.

sharonyang committed:
Added subclasses and rudimentary functions in statistics.py.

This should not yet be runnable because we do not have the Flow
class yet. Many of the implementation approaches need to be
reviewed.

lakshbhasin commented on network_simulator/__init__.py:
```diff
@@ -1,7 +1,9 @@
-"""Initialization file for the package network_simulator.
+"""
+Initialization file for the package network_simulator.
```
It would be useful to clarify that initializer.py is the actual entry-point into the simulation. I don't think we'll be using __init__.py at all. In any case, if we want to use this for PyPI, the submodules should be defined as shown here: https://docs.python.org/2/tutorial/modules.html#importing-from-a-package. Otherwise, "from network_simulator import *" does not work.

sharonyang replied:
Added a comment. Maybe removal in a later commit.

lakshbhasin commented on network_simulator/link.py:
```diff
@@ -17,8 +17,8 @@ class Link(object):
     def __init__(self, end_1_addr=None, end_2_addr=None, end_1_device=None,
-                 end_2_device=None, buffer=None, static_delay_sec=None,
-                 capacity_bps=None):
+                 end_2_device=None, buffer=None, buffer_occupancy=0,
```
The changes to link.py should wait until the PR #4 is merged in, since that makes a lot of substantial changes to link.py and this constructor.

With the new link.py, you will be able to find the buffer_occupancy by just finding the qsize() of the LinkBuffer's queue (qsize() is fine to use since we don't have multi-threading happening). Having an additional variable like this is not a good idea.

sharonyang replied:
Done.

lakshbhasin commented on network_simulator/link.py:
```diff
@@ -35,6 +35,8 @@ def __init__(self, end_1_addr=None, end_2_addr=None, end_1_device=None,
         self.end_2_device = end_2_device
         self.buffer = buffer
+        # TODO(team): Can we make sure if this can be implemented.
+        self.buffer_occupancy = buffer_occupancy
```
Remove as per above comment, and maybe wait till PR #4 is merged in.

sharonyang replied:
Done.

lakshbhasin commented on network_simulator/statistics.py:
```diff
@@ -4,6 +4,186 @@
+    :ivar list packet_transmit_times: a list of timestamps corresponding
+    to when a packet was transmitted.
+    """
+    def __init__(self, buffer_occupancy=0, packet_loss_times=[],
```
buffer_occupancy is a list of (timestamp, buffer_occupancy) tuples, so it should default to [].

sharonyang replied:
Done.

lakshbhasin commented on network_simulator/statistics.py:
```diff
@@ -4,6 +4,186 @@
+        self.packet_loss_times = packet_loss_times
+        self.packet_transmit_times = packet_transmit_times
+
+    # TODO(sharon): functions and subclasses/events (if any) goes here.
```
No Events are directly related to LinkStats, and there are no subclasses of LinkStats. Events can access LinkStats though. There is no need for additional functions imo, based on what you have below.

sharonyang replied:
Done.

lakshbhasin commented on network_simulator/statistics.py:
```diff
@@ -4,6 +4,186 @@
+        packet_rec_times_sec = packet_rec_times_sec
+        packet_rtts_sec = packet_rtts_sec
+
+    # TODO(sharon): functions and subclasses/events (if any) goes here.
```
Ditto what I wrote for LinkStats.

sharonyang replied:
Done.

lakshbhasin commented on network_simulator/statistics.py:
```diff
@@ -4,6 +4,186 @@
+
+class LinkStats(object):
+    """
+    Contains information of a link regarding to per-link buffer occupancy,
```
nit: "regarding to" --> ", related to". Same below.

sharonyang replied:
Done.

sharonyang replied:
Done.

lakshbhasin commented on network_simulator/statistics.py:
```diff
@@ -4,6 +4,186 @@
+    :ivar list packet_sent_times_sec: a list of timestamps corresponding to
+    when a data packet was sent.
+    :ivar list packet_rec_times_sec: a list of timestamps corresponding to
+    when an ACK packet was received for a given :class:`.Flow`.
```
nit: "for a given" --> "for this"

sharonyang replied:
Done.

lakshbhasin commented on network_simulator/statistics.py:
```diff
@@ -4,6 +4,186 @@
+        packet_sent_times_sec = packet_sent_times_sec
+        packet_rec_times_sec = packet_rec_times_sec
+
+    # TODO(sharon): functions and subclasses/events (if any) goes here.
```
Ditto what I wrote for LinkStats and FlowStats.

sharonyang replied:
Done.

lakshbhasin commented on network_simulator/statistics.py:
```diff
@@ -4,6 +4,186 @@
+    """
+     Contains information of a host regarding to per-host send/receive rate.
+
+    :ivar list packet_sent_times_sec: a list of (timestamp, packet_size)
```
Perhaps mention the packet size is in bits, to be clear.

sharonyang replied:
Done.

lakshbhasin commented on network_simulator/statistics.py:
```diff
@@ -4,6 +4,186 @@
+    Intended to be owned by the main loop to record all statistics within
+    the network.
+
+    :ivar dict link_stats: a map from (end_1_id, end_2_id) to
```
Perhaps mention that end_1_id, end_2_id, flow_id, and host_id are all strings.

sharonyang replied:
Done.

lakshbhasin commented on network_simulator/statistics.py:
```diff
@@ -4,6 +4,186 @@
+        skip the rest of the steps and return the stats object. If not,
+        create a new mapping then return the stats object.
+
+        :param Link link: link to be determined.
```
:return: LinkStats for that link. Same for functions below.

sharonyang replied:
Done.

lakshbhasin commented on network_simulator/statistics.py:
```diff
@@ -4,6 +4,186 @@
+        tup = (link.end_1_addr, link.end_2_addr)
+
+        if tup not in self.link_stats:
+            self.link_stats[tup] = LinkStats
```
Should be LinkStats() since an instance is being constructed. Same for FlowStats and HostStats below.

sharonyang replied:
Omg how did i do this. Done.

lakshbhasin commented on network_simulator/statistics.py:
```diff
@@ -4,6 +4,186 @@
+        :param Flow flow: flow to be determined.
+        """
+        if flow.id not in self.flow_stats:
+            self.flow_stats[flow.id] = FlowStats
```
I renamed the "id" attribute in flow.py to "flow_id", since "id" shadows an in-built variable name in Python. Might want to change this everywhere else that flow.id is used.

sharonyang replied:
Done.

lakshbhasin commented on network_simulator/statistics.py:
```diff
@@ -4,6 +4,186 @@
+        """
+        stats = self.get_link_stats(link)
+        # TODO(team): Is this the most efficient?
+        stats.buffer_occupancy.append((curr_time, link.buffer_occupancy))
```
The buffer occupancy should not be recorded here. Buffer occupancy changes should be recorded in a separate function, which will be called _whenever_ the link's buffer occupancy changes (as packets arrive and are sent off, _not_ just during losses). Define a separate function like:

```
def link_buffer_occ_change(self, link, curr_time):
    stats = self.get_link_stats(link)
    buffer_occ_packets = link.buffer.qsize()
    stats.buffer_occupancy.append((curr_time, buffer_occ_packets))
```

I believe that all packets (ACKs, data, and router) are included in "buffer occupancy", so qsize() should suffice. Might want to put in a TODO to clarify with Cody whether that's the case.

sharonyang replied:
Done.

lakshbhasin commented on network_simulator/statistics.py:
```diff
@@ -4,6 +4,186 @@
+        :param curr_time: simulation time of transmission.
+        """
+        stats = self.get_link_stats(link)
+        stats.buffer_occupancy.append((curr_time, link.buffer_occupancy))
```
Same as before, buffer occupancy changes should be recorded in a separate function, which will be called whenever the buffer size changes. The `link_packet_transmitted' function will only record transmission times and packet sizes of packets that were transmitted.

sharonyang replied:
Done.

lakshbhasin commented on network_simulator/statistics.py:
```diff
@@ -4,6 +4,186 @@
+    tuples.
+    :ivar list packet_loss_times: a list of timestamps, each corresponding
+    to a packet loss.
+    :ivar list packet_transmit_times: a list of timestamps corresponding
```
I just realized this needs to be a list of (timestamp, packet_size) tuples, where the packet_size is in bits. Otherwise, we cannot calculate the flow rate based on just the timestamps.

sharonyang replied:
Done.

lakshbhasin commented on network_simulator/statistics.py:
```diff
@@ -4,6 +4,186 @@
+
+class FlowStats(object):
+    """
+    Contains information of a flow regarding to per-flow send/receive rate
```
Technically, we just store timestamps here. The rates can be calculated since each sent packet has a fixed size (that of data packets), and each received packet has a fixed size (that of ACKs).

sharonyang replied:
Done.

lakshbhasin commented on network_simulator/statistics.py:
```diff
@@ -4,6 +4,186 @@
+    :ivar list packet_rec_times_sec: a list of (timestamp, packet_size)
+    tuples.
+    """
+    def __init__(self, packet_sent_times_sec=[], packet_rec_times_sec=[]):
```
Might be a good idea to remove the _sec from the variable names, and just specify that the timestamps are in seconds and the packet sizes are in bits in the :ivar documentation above.

sharonyang replied:
Done.

lakshbhasin commented on network_simulator/statistics.py:
```diff
@@ -4,6 +4,186 @@
+        sent_time = stats.packet_sent_times_sec[-1]
+        stats.packet_rtts_sec.append(curr_time - sent_time)
+
+    def host_packet_sent(self, host, curr_time):
```
packet_sent_times is a list of (timestamp, packet_size) tuples. So you need to pass in the packet size as a variable in this function, and include that in the packet_sent_times list when you append.

sharonyang replied:
Done.

lakshbhasin commented on network_simulator/statistics.py:
```diff
@@ -4,6 +4,186 @@
+        stats = self.get_host_stats(host)
+        stats.packet_sent_times_sec.append(curr_time)
+
+    def host_packet_received(self, host, curr_time):
```
Ditto what I mentioned above for host_packet_sent().

sharonyang replied:
Done.

sharonyang committed:
Merge branch 'master' of https://github.com/lakshbhasin/network_simulator into statistics-sharon

Conflicts:
	network_simulator/link.py

sharonyang committed:
Updated per Laksh's comments at PR #6.

This includes fixing bugs, adjusting some storage format, and updating
comments.

sharonyang committed:
Added type comments in statistics.py.

lakshbhasin commented on network_simulator/statistics.py:
```diff
@@ -4,6 +4,200 @@
+    :ivar list packet_rec_times: a list of timestamps corresponding to
+    when an ACK packet was received for this :class:`.Flow`. Time is in
+    seconds.
+    :ivar list packet_rtts: a list of RTTs (in seconds). Time is in
```
nit: "Time is in seconds" is redundant here.

sharonyang replied:
Done.

lakshbhasin commented on network_simulator/statistics.py:
```diff
@@ -4,6 +4,200 @@
+        :param Link link: link that has a change of buffer size.
+        """
+        stats = self.get_link_stats(link)
+        # TODO(sharon): Check with Cody if qsize() would work in this case.
```
Regarding the TODO: What we need to check is whether we want to count RoutingPackets as well (instead of just AckPackets and DataPackets), not if qsize() will work. The queue by default contains all kinds of packets at the moment (we might put RoutingPackets somewhere else later or have them bypass the queue).

sharonyang replied:
Ok. Modified TODO.

lakshbhasin commented on network_simulator/statistics.py:
```diff
@@ -4,6 +4,200 @@
+        """
+        stats = self.get_link_stats(link)
+        # TODO(sharon): Check with Cody if qsize() would work in this case.
+        buffer_occ_packets = link.buffer.qsize()
```
link.link_buffer.qsize()

sharonyang replied:
Done.

lakshbhasin commented on network_simulator/statistics.py:
```diff
@@ -4,6 +4,200 @@
+
+    def flow_packet_sent(self, flow, curr_time):
+        """
+        Record a packet sent from a flow.
```
Clarify this is a DataPacket being sent.

sharonyang replied:
Done.

lakshbhasin commented on network_simulator/statistics.py:
```diff
@@ -4,6 +4,200 @@
+
+    def flow_packet_received(self, flow, curr_time):
+        """
+        Record a packet received by a flow.
```
Clarify this is an AckPacket being received.

sharonyang replied:
Done.

lakshbhasin commented on network_simulator/statistics.py:
```diff
@@ -4,6 +4,200 @@
+        stats = self.get_flow_stats(flow)
+        stats.packet_sent_times.append(curr_time)
+
+    def flow_packet_received(self, flow, curr_time):
```
You should pass in the AckPacket itself, as well as curr_time (unlike flow_packet_sent).

sharonyang replied:
Done.

lakshbhasin commented on network_simulator/statistics.py:
```diff
@@ -4,6 +4,200 @@
+        stats = self.get_link_stats(link)
+        stats.packet_transmit_times.append((curr_time, packet_size))
+
+    def flow_packet_sent(self, flow, curr_time):
```
Might be cleaner to pass in the DataPacket itself here, and just read off its start_time_sec, instead of passing in curr_time.

sharonyang replied:
Done.

lakshbhasin commented on network_simulator/statistics.py:
```diff
@@ -4,6 +4,200 @@
+        # Retrieve the last element on the sent_times list to find
+        # the corresponding sent time and using the diff we can
+        # determine the rtt time.
+        sent_time = stats.packet_sent_times[-1]
```
This is a buggy way to calculate RTT, since we might send a packet soon before receiving back an unrelated packet.

Instead, we should just pass in the AckPacket into this function. Add a data_packet_start_time_sec field to AckPacket so that the AckPacket can track the original DataPacket's start time ("original DataPacket" = the one that triggered the ACK from the Host). Then, the RTT is just curr_time - ack_packet.data_packet_start_time_sec.

sharonyang replied:
Done.

lakshbhasin commented on network_simulator/statistics.py:
```diff
@@ -4,6 +4,200 @@
+        sent_time = stats.packet_sent_times[-1]
+        stats.packet_rtts.append(curr_time - sent_time)
+
+    def host_packet_sent(self, host, packet_size, curr_time):
```
Again, you can pass in the packet itself instead of packet_size and curr_time, and then the packet_sent_times element will depend on the Packet's start_time_sec and the Packet's size_bits. This is probably cleaner.

sharonyang replied:
Done.

lakshbhasin commented on network_simulator/statistics.py:
```diff
@@ -4,6 +4,200 @@
+        stats = self.get_host_stats(host)
+        stats.packet_sent_times.append((curr_time, packet_size))
+
+    def host_packet_received(self, host, curr_time):
```
Pass in the packet itself since you need its size. This time, you should leave in curr_time of course.

sharonyang replied:
Done.

lakshbhasin commented on network_simulator/statistics.py:
```diff
@@ -4,6 +4,200 @@
+        :param float curr_time: simulation time of reception.
+        """
+        stats = self.get_host_stats(host)
+        stats.packet_rec_times.append(curr_time)
```
packet_rec_times is a list of (timestamp, packet_size) tuples.

sharonyang replied:
Done.

sharonyang committed:
Modified based on Laksh's comment 2 from PR #6.

This includes updating parameters of statistics.py functions and
their comments.

lakshbhasin commented on network_simulator/packet.py:
```diff
@@ -65,13 +65,15 @@ class AckPacket(Packet):
     """
```
Add ivar with data_packet_start_time_sec to AckPacket documentation.

sharonyang replied:
Done.

lakshbhasin commented on network_simulator/statistics.py:
```diff
@@ -4,6 +4,203 @@
+        stats = self.get_flow_stats(flow)
+        stats.packet_rec_times.append(curr_time)
+        # Retrieve data packet sent time from the ack packet and use
+        # it to calculate RTT time.
```
nit: "RTT time" is redundant.

sharonyang replied:
Done.

lakshbhasin commented:
Looks good to merge after you address the two remaining comments. FYI the next statistics PR should handle graphing time traces of everything in matplotlib (somehow we'll have to do this with lots of plots and make it look OK), as well as calculating averages.

sharonyang committed:
Fixes after Laksh's comment 3 for PR #6.

This includes updating comments for statistics.py and packet.py.
